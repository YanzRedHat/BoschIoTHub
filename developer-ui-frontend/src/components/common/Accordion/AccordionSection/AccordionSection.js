/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React, { Children, cloneElement } from "react";
import PropTypes from "prop-types";
import requiredIf from "react-required-if";

const AccordionSection = ({ children, className, ...other }) => (
  <div className={className ? className : null}>
    {/* Spread the props to all children */}
    {Children.map(children, child => cloneElement(child, { ...other }))}
  </div>
);

AccordionSection.propTypes = {
  expanded: requiredIf(PropTypes.bool, props => !props.sticky),
  toggleExpanded: requiredIf(PropTypes.func, props => !props.sticky),
  sticky: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  children: props => {
    if (
      !Array.isArray(props.children) ||
      props.children.length !== 2 ||
      !props.children.every(React.isValidElement)
    ) {
      return new Error(
        "children needs to be an array of two Elements (AccordionSectionHeader and AccordionSectionBody)"
      );
    }
    return null;
  }
};

export default AccordionSection;
