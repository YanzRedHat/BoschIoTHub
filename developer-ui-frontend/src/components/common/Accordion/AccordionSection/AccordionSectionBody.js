/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React, { Component } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import Measure from "react-measure";
import { Motion, spring } from "react-motion";

const AnimationWrapper = styled.div`
  overflow: hidden;
`;

export default class AccordionSectionBody extends Component {
  constructor(props) {
    super(props);
    this.state = {
      height: -1
    };
  }

  render() {
    const { expanded, children, sticky } = this.props;
    const { height } = this.state;
    return !sticky ? (
      <Measure
        scroll
        onResize={contentRect => {
          this.setState({ height: contentRect.scroll.height });
        }}>
        {({ measureRef }) => (
          <Motion
            style={{
              height: spring(expanded ? height : 0, {
                stiffness: 300,
                damping: 27
              })
            }}
            defaultStyle={{ height: 0 }}>
            {interpolatingStyle => (
              <AnimationWrapper style={interpolatingStyle}>
                <div className="measuring-container" ref={measureRef}>
                  {children}
                </div>
              </AnimationWrapper>
            )}
          </Motion>
        )}
      </Measure>
    ) : (
      children
    );
  }
}

AccordionSectionBody.propTypes = {
  expanded: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  sticky: PropTypes.bool
};
