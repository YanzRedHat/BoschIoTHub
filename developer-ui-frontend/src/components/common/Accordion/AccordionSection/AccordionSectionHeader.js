/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
// SVG Imports
import CaretDown from "images/caretDownIcon.svg";

const ClickProxy = styled.span`
  ${props => !props.sticky && "cursor: pointer"};
  display: inline-flex;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  z-index: 1;
  align-items: center;
  white-space: nowrap;
  overflow: hidden;
`;

const HeaderContainer = styled.div`
  position: relative;
  z-index: 0;
  display: flex;
  justify-content: space-between;
  height: 4rem;
  align-content: center;
  background: #fff;
  border: 1px solid #ccc;
  font-weight: 500;
  padding: 0 1rem;
  color: #757575;
  transition: box-shadow 0.4s ease-out;
  ${props =>
    props.expanded &&
    `box-shadow: inset 0px -5px 5px 0px rgba(28, 33, 35, 0.14);`};
  ${props =>
    props.disabled &&
    `
    pointer-events: none;
    background: #e4e4e4 !important;
    color: rgba(33, 33, 33, 0.47) !important;
  `}
`;

const ActionButtonsWrapper = styled.span`
  display: inline-flex;
  z-index: 2;
  overflow: visible;
  align-items: center;
  white-space: nowrap;

  svg {
    cursor: pointer;
    width: 1.5rem;
    height: 1.5rem;

    path {
      fill: #757575;
    }
  }
`;

const TitleContainer = styled.span`
  display: inline-flex;
  align-items: center;
  white-space: nowrap;
  overflow: hidden;
`;

const TitleIcon = styled(({ icon, ...props }) =>
  React.cloneElement(icon, props)
)`
  display: inline-block;
  height: 2rem;
  width: 2rem;
  margin-right: 1rem;
  overflow: visible;
  path {
    fill: ${props => {
      if (props.disabled) {
        return "rgba(33, 33, 33, 0.47)";
      }
      return props.theme.accentColor;
    }};
  }
`;

const AccordionSectionHeader = ({
  title,
  icon,
  expanded,
  toggleExpanded,
  children,
  sticky,
  disabled,
  ...other
}) => (
  <div>
    <HeaderContainer expanded={expanded} disabled={disabled} {...other}>
      <ClickProxy onClick={toggleExpanded} sticky={sticky} />
      <TitleContainer>
        {icon ? <TitleIcon icon={icon} disabled={disabled} /> : null}
        {title}
      </TitleContainer>
      <ActionButtonsWrapper>
        {(expanded || sticky) && children}
        {!sticky && !disabled && (
          <CaretDown
            className={
              expanded ? "header-icon-right caret-rotated" : "header-icon-right"
            }
            onClick={toggleExpanded}
          />
        )}
      </ActionButtonsWrapper>
    </HeaderContainer>
  </div>
);

AccordionSectionHeader.propTypes = {
  expanded: PropTypes.bool,
  disabled: PropTypes.bool,
  toggleExpanded: PropTypes.func,
  title: PropTypes.string.isRequired,
  sticky: PropTypes.bool,
  icon: PropTypes.element,
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ])
};

export default AccordionSectionHeader;
