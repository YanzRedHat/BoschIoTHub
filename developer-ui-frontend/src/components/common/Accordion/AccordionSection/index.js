/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
export { default } from "./AccordionSection";
export { default as AccordionSectionHeader } from "./AccordionSectionHeader";
export { default as AccordionSectionBody } from "./AccordionSectionBody";
