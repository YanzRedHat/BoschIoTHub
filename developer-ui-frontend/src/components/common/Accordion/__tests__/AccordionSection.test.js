/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React from "react";
import { shallow } from "enzyme";
import AccordionSection from "components/common/Accordion/AccordionSection/AccordionSection.js";
import AccordionSectionHeader from "components/common/Accordion/AccordionSection/AccordionSectionHeader.js";
import AccordionSectionBody from "components/common/Accordion/AccordionSection/AccordionSectionBody.js";

describe("AccordionSection", () => {
  it("should be defined", () => {
    expect(AccordionSection).toBeDefined();
  });
  it("should render correctly", () => {
    const tree = shallow(
      <AccordionSection name="accordionSection test" expanded />
    );
    expect(tree).toMatchSnapshot();
  });
  it.skip("should have a prop className", () => {
    const tree = shallow(
      <AccordionSection name="accordionSection test" className="test" />
    );
    expect(tree.props().className).toEqual("test");
  });
  it.skip("should have a prop children", () => {
    const tree = shallow(
      <AccordionSection
        name="accordionSection test"
        children={[AccordionSectionHeader, AccordionSectionBody]}
      />
    );
    expect(tree.props().children).toEqual([
      AccordionSectionHeader,
      AccordionSectionBody
    ]);
  });
});
