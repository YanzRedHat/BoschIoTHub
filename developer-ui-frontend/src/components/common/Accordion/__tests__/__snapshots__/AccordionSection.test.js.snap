// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionSection should render correctly 1`] = `
<div
  className={null}
/>
`;
